llama-index-core/llama_index/core/extractors/document_context.py

```
@@ -0,0 +1,331 @@
+        key (str): Metadata key for storing extracted context
+        prompt (str): Prompt template for context generation
+        doc_ids (Set[str]): Set of processed document IDs
+        max_context_length (int): Maximum allowed document context length
+        max_output_tokens (int): Maximum tokens in generated context
+        oversized_document_strategy (OversizeStrategy): Strategy for handling large documents
+
+    Example:
+        ```python
+        extractor = DocumentContextExtractor(
+            docstore=my_docstore,
+            llm=my_llm,
+            max_context_length=64000,
+            max_output_tokens=256
+        )
+        metadata_list = await extractor.aextract(nodes)
+        ```
+    """
+
+    # Pydantic fields
+    llm: LLM
+    docstore: DocumentStore
+    key: str
+    prompt: str
+    doc_ids: Set[str]
+    max_context_length: int
+    max_output_tokens: int
+    oversized_document_strategy: OversizeStrategy
+    num_workers: int = DEFAULT_NUM_WORKERS
+    failed_nodes: Set[str]
+
+    ORIGINAL_CONTEXT_PROMPT: ClassVar[str] = ORIGINAL_CONTEXT_PROMPT
+    SUCCINCT_CONTEXT_PROMPT: ClassVar[str] = SUCCINCT_CONTEXT_PROMPT
+
+    DEFAULT_KEY: str = "context"
+
+    def is_job_complete(self) -> bool:
+        """
+        Simply tracks if the contextual retrieval is 100% finished through all nodes.
+        """
+        return len(self.failed_nodes) == 0
+
+    def __init__(
+        self,
+        docstore: DocumentStore,
+        llm: Optional[LLM] = None,
+        max_context_length: int = 1000,
+        key: str = DEFAULT_KEY,
+        prompt: str = ORIGINAL_CONTEXT_PROMPT,
+        num_workers: int = DEFAULT_NUM_WORKERS,
+        max_output_tokens: int = 512,
+        oversized_document_strategy: OversizeStrategy = "warn",
+        **kwargs: Any,
+    ) -> None:
+        """Init params."""
+        assert hasattr(
+            llm, "achat"
+        )  # not all LLMs have this, particularly the huggingfaceapi ones.
```

every LLM will have this attribute, but it may or may not be implemented 🤔 (i.e. you'll only know if you call it)

---

ah, crap! well what am I supposed to do about that? :D 
I guess I need a try except and a synchronous fallback?

---

I think if achat isn't implemented, we should just let it raise -- falling back to sync will be extremely slow for users 🤔